src/hydro_forces.cpp

```diff
@@ -353,58 +353,48 @@ double TestHydro::setVelHistory(double val, int step, int b_num, int index) {
+		auto gg = body->GetSystem()->Get_G_acc().Length();
+
+		// hydrostatic stiffness due to offset from equilibrium
+		auto body_position = body->GetPos();
+		auto body_rotation = body->GetRot().Q_to_Euler123();
+		// calculate displacement
+		chrono::ChVectorN<double, 6> displacement;
+		for (int ii = 0; ii < 3; ii++) {
+			displacement[ii] = body_position[ii] - equilibrium[ii];
+			displacement[ii + 3] = body_rotation[ii] - equilibrium[ii + 3];
 		}
-	}
-
-	// make displacement vector for system
-	for (int i = 0; i < total_dofs; i++) {
-		displacement[i] = equilibrium[i] - position[i];
-	}
-
-	// re invent matrix vector multiplication
-	for (int b = 0; b < num_bodies; b++) {
-		unsigned b_offset = 6 * b;
-		for (int i = 0; i < 6; i++) {
-			for (int j = 0; j < 6; j++) {
-				double t = (file_info[b].GetHydrostaticStiffness(i, j));
-				//if (i == 4 && j == 4 && t < 0) {
-				//	t = -t;
-				//}
-				force_hydrostatic[i + b_offset] += t * displacement[j + b_offset];
-			}
+		// calculate force
+		auto force_offset = -gg * rho * body_h5file.lin_matrix * displacement;
+		// add to force_hydrostatic
+		for (int dof = 0; dof < 6; dof++) {
+			body_force_hydrostatic[dof] += force_offset[dof];
 		}
-	}
 
-
-	// now handle buoyancy force....
-	assert(num_bodies > 0);
-	double* buoyancy = new double[num_bodies]; // this sets up an array for buoyancy for each body
-	// add heave buoyancy for each body, and add rxb=(cb-cg)x(0,0,buoyancy) for the moment due to buoyancy for each body (simplified)
-	for (int b = 0; b < num_bodies; b++) {
-		buoyancy[b] = file_info[b].rho * -(bodies[b]->GetSystem()->Get_G_acc()).z() * file_info[b].disp_vol; // buoyancy = rho*g*Vdisp
-		unsigned b_offset = 6 * b; // force_hydrostatic has 6 elements for each body so to skip to the next body we move 6 spaces
-		unsigned r_offset = 3 * b; // cb_minus_cg has 3 elements for each body so to skip to the next body we move 3 spaces
-		force_hydrostatic[b_offset + 2] += buoyancy[b]; // add heave buoyancy
-		// now for moments due to buoyancy (simplified)
-		// for torque about x (index 3) per body, add b * r_y
-		force_hydrostatic[b_offset + 3] += buoyancy[b] * cb_minus_cg[1 + r_offset];
-		// for torque about y (index 4) per body, subtract b*r_x
-		force_hydrostatic[b_offset + 4] += -1 * buoyancy[b] * cb_minus_cg[0 + r_offset];
+		// buoyancy at equilibrium
+		// TODO: move to prestep (shouldn't be calculated at each time step)
+		// translational
+		auto buoyancy = rho * (-bodies[b]->GetSystem()->Get_G_acc()) * body_h5file.disp_vol; // buoyancy = rho*g*Vdisp
+		body_force_hydrostatic[0] += buoyancy[0];
+		body_force_hydrostatic[1] += buoyancy[1];
+		body_force_hydrostatic[2] += buoyancy[2];
+		// rotational
+		unsigned r_offset = 3 * b;  // cb_minus_cg has 3 elements for each body so to skip to the next body we move 3 spaces
+		auto cg2cb = chrono::ChVector<double>(cb_minus_cg[0 + r_offset], cb_minus_cg[0 + r_offset], cb_minus_cg[0 + r_offset]);
```

Comment:
I think the indexing is wrong here? I'm not sure but I think it should look something like this 
```suggestion
		chrono::ChVector<double> cg2cb = (cb_minus_cg[0 + r_offset], cb_minus_cg[1 + r_offset], cb_minus_cg[2 + r_offset]);
```

Reply:
Good catch! Yes it should be 0, 1, 2